systems-manager.tf

```diff
@@ -0,0 +1,98 @@
+# AWS KMS alias used for encryption/decryption of SSM secure strings
+variable "kms_alias_name_ssm" {
+  type        = string
+  default     = "alias/aws/ssm"
+  description = "KMS alias name for SSM"
+}
+
+variable "ssm_enabled" {
+  type        = bool
+  default     = false
+  description = "If `true` create SSM keys for the database user and password."
+}
+
+variable "ssm_key_format" {
+  type        = string
+  default     = "/%v/%v/%v"
+  description = "SSM path format. The values will will be used in the following order: `var.ssm_key_prefix`, `var.database_name`, `var.ssm_key_*`"
+}
+
+variable "ssm_key_prefix" {
+  type        = string
+  default     = "rds"
+  description = "SSM path prefix. Omit the leading forward slash `/`."
+}
+
+variable "ssm_key_user" {
+  type        = string
+  default     = "admin/db_user"
+  description = "The SSM key to save the user. See `var.ssm_path_format`."
+}
+
+variable "ssm_key_password" {
+  type        = string
+  default     = "admin/db_password"
+  description = "The SSM key to save the password. See `var.ssm_path_format`."
+}
+
+variable "ssm_key_hostname" {
+  type        = string
+  default     = "admin/db_hostname"
+  description = "The SSM key to save the hostname. See `var.ssm_path_format`."
+}
+
+variable "ssm_key_port" {
+  type        = string
+  default     = "admin/db_port"
+  description = "The SSM key to save the port. See `var.ssm_path_format`."
+}
+
+locals {
+  ssm_enabled = local.enabled && var.ssm_enabled
+}
+
+resource "aws_ssm_parameter" "rds_database_user" {
+  count = local.ssm_enabled ? 1 : 0
+
+  name        = format(var.ssm_key_format, var.ssm_key_prefix, var.database_name, var.ssm_key_user)
+  value       = local.database_user
+  description = "RDS DB user"
+  type        = "String"
+  overwrite   = true
+}
+
+resource "aws_ssm_parameter" "rds_database_password" {
+  count = local.ssm_enabled ? 1 : 0
+
+  name        = format(var.ssm_key_format, var.ssm_key_prefix, var.database_name, var.ssm_key_password)
+  value       = local.database_password
+  description = "RDS DB password"
+  type        = "SecureString"
+  key_id      = var.kms_alias_name_ssm
+  overwrite   = true
+}
+
+resource "aws_ssm_parameter" "rds_database_hostname" {
+  count = local.ssm_enabled ? 1 : 0
+
+  name        = format(var.ssm_key_format, var.ssm_key_prefix, var.database_name, var.ssm_key_hostname)
+  value       = module.dns_host_name.hostname
+  description = "RDS DB hostname"
+  type        = "String"
+  overwrite   = true
+}
+
+resource "aws_ssm_parameter" "rds_database_port" {
```

Comment:
we should have a secrets module where we can pass a map of secrets and so people can chose to use SSM or secret manager etc

Reply:
@jamengual just randomly spotting this while scrolling the onslaught of emails and just wanted to chime in: I think that's a great call. If ya'll want to discuss a solution to figuring out an agnostic way to deal with SSM PStore, Secrets Manager, AND Sops files then that'd be something I'd be very keen on helping out with. 

cc @osterman

Reply:
@jamengual that sounds like a great idea.

@Gowiem @osterman is creating a separate SSM module a blocker for this ?

Perhaps

```hcl
variable "save_parameters_to_ssm_enabled" { default = true }
variable "save_parameters_to_secrets_manager_enabled" { default = false }
variable "ssm_key_format" { default = "/%v/%v/%v" }
variable "ssm_key_prefix" { default = "rds" }
variable "save_parameter_ssm_map" {
  default = {
    user = {
      suffix      = "admin/db_user"
      description = "RDS DB user"
      type        = "String"
      overwrite   = true
    }
    password = {
      suffix      = "admin/db_password"
      description = "RDS DB password"
      type        = "SecureString"
      overwrite   = true
    }
    hostname = {
      suffix      = "admin/db_hostname"
      description = "RDS DB hostname"
      type        = "String"
      overwrite   = true
    }
    port = {
      suffix      = "admin/db_port"
      description = "RDS DB port"
      type        = "String"
      overwrite   = true
    }
  }
}

locals {
  value_parameter_map = {
    user     = local.database_user
    password = local.database_password
    hostname = module.dns_host_name.hostname
    port     = var.database_port
  }
}

resource "aws_ssm_parameter" "rds_database" {
  for_each = var.save_parameter_ssm_map

  name        = format(var.ssm_key_format, var.ssm_key_prefix, var.database_name, each.value.suffix)
  value       = local.value_parameter_map[each.key]
  description = each.value.description
  type        = each.value.type
  key_id      = each.value.type == "SecureString" ? var.kms_alias_name_ssm : null
  overwrite   = each.value.overwrite
}
```

Reply:
Only thing is that `secretsmanager` is meant for secrets and only the `password` parameter is a real secret that needs to be encrypted.

Reply:
I used the above map but did not incorporate the `secretsmanager` for now.